### DISCUSSION TITLE
Poor storj performance

### DISCUSSION THREAD

**Discussion by Villain88** [gramineproject/gramine#1471] [General]
Hello.

We use the [uplink library](https://github.com/storj/uplink-c) to interact with[ Storj](https://www.storj.io/).
We are experiencing performance issues with file uploads.
Outside the enclave, there are no such problems, but if using gramine-direct, the issue is still present.
If described briefly, how Storj works, is that it connects to a multitude of nodes for file uploads.
During debugging, it was found that the slow file upload is due to the extended connection time to the nodes or attempting to connect to an unavailable node. 
The connection can last for 2 minutes until the system call 'connect' times out.

I have 2 assumptions why this might be happening:

1. gramine-sgx/gramine-direct cannot determine the remoteness and availability of hosts, and then uplink attempts to connect to random hosts that may be unreachable or have high latency.
2. Uplink is written in GoLang (there is only a wrapper for C) and utilizes goroutines. There might be some issues with their interpretation in gramine.

I have written a small example for demonstration:
https://github.com/Villain88/storj-example

_The performance result without the enclave:_
1 - 65536 bytes successfully uploaded in 1538ms
2 - 65536 bytes successfully uploaded in 1445ms
3 - 65536 bytes successfully uploaded in 1408ms
4 - 65536 bytes successfully uploaded in 1406ms
5 - 65536 bytes successfully uploaded in 1336ms 

_In enclave:_
1 - 65536 bytes successfully uploaded in 9018ms
2 - 65536 bytes successfully uploaded in 8066ms
3 - 65536 bytes successfully uploaded in 16395ms
4 - 65536 bytes successfully uploaded in **131723**ms
5 - 65536 bytes successfully uploaded in **132154**ms 

Inside the file **test.cpp**, there is an access token. You can use it for testing purposes or use your own if possible.
For security reasons, the token is available for 3 months, but if necessary, I will be able to extend the access."

Uplink is memory and thread intensive, and I would recommend using an enclave with a minimum of 8GB of memory and at least 256 threads   

I will be grateful for any assistance.

**Comment by dimakuv** [comment]
@Villain88 Could you add a Gramine manifest template file to your https://github.com/Villain88/storj-example example? Would be easier to reproduce, and maybe we'll be able to notice some bugs in the manifest template, which could solve your problem (without needing to reproduce on our systems).

@aneessahib Do we have anyone who could take this task? We should probably put it in our list of pending tasks.

**Reply by aneessahib** [reply]
Yes we should add to backlog,  - `storej` looks like a good usecase. 

>The performance result without the enclave:
1 - 65536 bytes successfully uploaded in 1538ms
2 - 65536 bytes successfully uploaded in 1445ms
3 - 65536 bytes successfully uploaded in 1408ms
4 - 65536 bytes successfully uploaded in 1406ms
5 - 65536 bytes successfully uploaded in 1336ms

Is this gramine-direct, or native execution numbers?

**Reply by Villain88** [reply]
native, gramine-direct also shows poor performance.

**Comment by Villain88** [comment]
> @Villain88 Could you add a Gramine manifest template file to your https://github.com/Villain88/storj-example example? Would be easier to reproduce, and maybe we'll be able to notice some bugs in the manifest template, which could solve your problem (without needing to reproduce on our systems).

I added a manifest and Dockerfile, and also modified the build script, as we are using gramenized containers.

**Comment by aneessahib** [comment]
@amathew3

**Reply by amathew3** [reply]
Storj uses non-blocking sockets to connect to its nodes. Gramine implementation of connect with non-blocking mode convert all non-blocking calls to blocking calls. 
If the connect returns EINPROGRESS, Gramine code wait for connection PASS/FAIL before returning to application code.  The following sequence is followed in the connect implementation.

`connect(..)
poll(fd)
getsockopt(..)`

This blocking behavior adds overhead to storj node connections.

**Reply by dimakuv** [reply]
@amathew3 Interesting finding. Can you provide the links to Gramine sources?

Also, what would be the fix for this? Can we just return `EINPROGRESS` to the application, instead of going into the loop? Do you have such a patch to Gramine, and did it show improvement for Storj?

**Reply by amathew3** [reply]
@dimakuv Here is the Linux PAL implementation link:https://github.com/gramineproject/gramine/blob/master/pal/src/host/linux/pal_sockets.c

I have a patch with similar changes, yet to test it. The access token expired in the test code given.

@Villain88 
Could you please  generate new access token, getting Unauthorized  API credentials error with the test app.
![image](https://github.com/gramineproject/gramine/assets/72502805/213abf42-2ca4-400a-8fca-12e60e2c436f)

**Reply by aneessahib** [reply]
@amathew3 I think its worth mentioning regarding the existing comment on this from Gramine sources. https://github.com/gramineproject/gramine/blob/master/pal/src/host/linux/pal_sockets.c#L279C28-L279C28

**Reply by Villain88** [reply]
@amathew3 
Updated the token in the [repository](https://github.com/Villain88/storj-example) 
It will be active until January 26, 2024

**Reply by amathew3** [reply]
@dimakuv I have tested the patch mentioned earlier, observed better performance for storj. Here is the performance comparison data.
![image](https://github.com/gramineproject/gramine/assets/72502805/6ce67539-e89a-45d0-bec3-40796c5dfed2)

**Reply by dimakuv** [reply]
@amathew3 Thanks! The data looks very good, please show us the patch at some point and we'll figure out how to proceed with it.
